[seahawk_rov.logic_tube_bno085] error linear_acceleration must be type Vector3 on imu publisher

ERROR
```
ubuntu@SeaHawk:~/cabrillo_rov_2023$ ros2 launch ./launch/rov.yaml 
[INFO] [launch]: All log files can be found below /home/ubuntu/.ros/log/2023-02-10-23-57-26-400634-SeaHawk-19981
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [h264_cam_node-1]: process started with pid [19982]
[INFO] [i2c_proxy-2]: process started with pid [19984]
[h264_cam_node-1] [INFO] [1676073447.458060022] [front_camera]: Parameter input_fn: /dev/video0
[h264_cam_node-1] [INFO] [1676073447.458894532] [front_camera]: Parameter fps: 20
[h264_cam_node-1] [INFO] [1676073447.458997697] [front_camera]: Parameter size: 800x600
[h264_cam_node-1] [INFO] [1676073447.459043974] [front_camera]: Parameter frame_id: rov_frame
[h264_cam_node-1] [INFO] [1676073447.459080604] [front_camera]: Parameter camera_info_path: info.ini
[h264_cam_node-1] [video4linux2,v4l2 @ 0xaaab11612180] Cannot open video device /dev/video0: No such file or directory
[h264_cam_node-1] [ERROR] [1676073447.459660802] [front_camera]: Could not open the v4l device: /dev/video0
[h264_cam_node-1] terminate called after throwing an instance of 'std::runtime_error'
[h264_cam_node-1]   what():  Could not open the v4l device
[ERROR] [h264_cam_node-1]: process has died [pid 19982, exit code -6, cmd '/home/ubuntu/cabrillo_rov_2023/install/h264_image_transport/lib/h264_image_transport/h264_cam_node --ros-args -r __node:=front_camera --params-file /tmp/launch_params_j8c6bvbe --params-file /tmp/launch_params_slqn7m32 --params-file /tmp/launch_params_ey3chm05 --params-file /tmp/launch_params_yt6bw64c -r image_raw/h264:=front_camera/h264'].
[i2c_proxy-2] Traceback (most recent call last):
[i2c_proxy-2]   File "/home/ubuntu/cabrillo_rov_2023/install/seahawk_rov/lib/seahawk_rov/i2c_proxy", line 33, in <module>
[i2c_proxy-2]     sys.exit(load_entry_point('seahawk-rov', 'console_scripts', 'i2c_proxy')())
[i2c_proxy-2]   File "/home/ubuntu/cabrillo_rov_2023/build/seahawk_rov/seahawk_rov/i2c_proxy/i2c_proxy.py", line 145, in main
[i2c_proxy-2]     rclpy.spin(node_i2c_proxy)
[i2c_proxy-2]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/__init__.py", line 222, in spin
[i2c_proxy-2]     executor.spin_once()
[i2c_proxy-2]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 713, in spin_once
[i2c_proxy-2]     raise handler.exception()
[i2c_proxy-2]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/task.py", line 239, in __call__
[i2c_proxy-2]     self._handler.send(None)
[i2c_proxy-2]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 418, in handler
[i2c_proxy-2]     await call_coroutine(entity, arg)
[i2c_proxy-2]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 332, in _execute_timer
[i2c_proxy-2]     await await_or_execute(tmr.callback)
[i2c_proxy-2]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 107, in await_or_execute
[i2c_proxy-2]     return callback(*args)
[i2c_proxy-2]   File "/home/ubuntu/cabrillo_rov_2023/build/seahawk_rov/seahawk_rov/i2c_proxy/i2c_proxy.py", line 127, in poll_sensors
[i2c_proxy-2]     message_logic_tube_imu.linear_acceleration = logic_tube_imu.raw_acceleration
[i2c_proxy-2]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/sensor_msgs/msg/_imu.py", line 297, in linear_acceleration
[i2c_proxy-2]     assert \
[i2c_proxy-2] AssertionError: The 'linear_acceleration' field must be a sub message of type 'Vector3'
[ERROR] [i2c_proxy-2]: process has died [pid 19984, exit code 1, cmd '/home/ubuntu/cabrillo_rov_2023/install/seahawk_rov/lib/seahawk_rov/i2c_proxy --ros-args -r __node:=i2c_proxy'].
ubuntu@SeaHawk:~/cabrillo_rov_2023$ 
```

i believe the important bit is this
```
[i2c_proxy-2] AssertionError: The 'linear_acceleration' field must be a sub message of type 'Vector3'
```

TODO:
find out what type the raw_acceleration is on the bno08x package.
transform that type to a vector3

closes #24

--- a/launch/rov.launch.py
+++ b/launch/rov.launch.py
@@ -24,5 +24,30 @@ def generate_launch_description():
             executable='seahawk_rov',
             name='seahawk_rov',
             output='screen'
+        ),
+        Node(
+            package='tf2_ros',
+            executable='static_transform_publisher',
+            name='imu_transform',
+            output='screen',
+            arguments=[
+                # Offset from parent frame in meters
+                "--x",      "0",
+                "--y",      "0",
+                "--z",      "0",
+                # # Quaternion rotation. Alternately use --{roll,pitch,yaw} but those probably aren't ideal in the end
+                # "--qx",     "0",
+                # "--qy",     "0",
+                # "--qz",     "0",
+                # "--qw",     "0",
+                # RPY rotation in radians. This is not ideal, it would be better to use Quaternion rotation
+                "--roll",   "0",
+                "--pitch",  "0",
+                "--yaw",    "0",
+                # Parent frame to which the offset is provided
+                "--frame-id", "base_link",
+                # Created frame which is offset from parent frame
+                "--child-frame-id", "logic_tube_bno085",
+            ]
         )
     ])

--- a/launch/rov.launch.py
+++ b/launch/rov.launch.py
@@ -1,7 +1,7 @@
 from launch import LaunchDescription
 from launch_ros.actions import Node
 
-from math import pi
+# from math import pi
 
 def generate_launch_description():
     return LaunchDescription([
@@ -41,10 +41,10 @@ def generate_launch_description():
                 # "--qy",     "0",
                 # "--qz",     "0",
                 # "--qw",     "0",
-                # RPY rotation in radians. This is not ideal, it would be better to use Quaternion rotation
-                "--roll",   f"{pi}",
-                "--pitch",  f"0",
-                "--yaw",    f"{-pi*0.5}",
+                # # RPY rotation in radians. This is not ideal, it would be better to use Quaternion rotation
+                # "--roll",   f"{pi}",
+                # "--pitch",  f"0",
+                # "--yaw",    f"{-pi*0.5}",
                 # Parent frame to which the offset is provided
                 "--frame-id", "base_link",
                 # Created frame which is offset from parent frame

--- a/launch/rov.launch.py
+++ b/launch/rov.launch.py
@@ -1,6 +1,7 @@
 from launch import LaunchDescription
 from launch_ros.actions import Node
 
+from math import pi
 
 def generate_launch_description():
     return LaunchDescription([
@@ -32,18 +33,18 @@ def generate_launch_description():
             output='screen',
             arguments=[
                 # Offset from parent frame in meters
-                "--x",      "0",
-                "--y",      "0",
-                "--z",      "0",
+                "--x",      "0.17",
+                "--y",      "0.05",
+                "--z",      "0.07",
                 # # Quaternion rotation. Alternately use --{roll,pitch,yaw} but those probably aren't ideal in the end
                 # "--qx",     "0",
                 # "--qy",     "0",
                 # "--qz",     "0",
                 # "--qw",     "0",
                 # RPY rotation in radians. This is not ideal, it would be better to use Quaternion rotation
-                "--roll",   "0",
-                "--pitch",  "0",
-                "--yaw",    "0",
+                "--roll",   f"{pi}",
+                "--pitch",  f"0",
+                "--yaw",    f"{-pi*0.5}",
                 # Parent frame to which the offset is provided
                 "--frame-id", "base_link",
                 # Created frame which is offset from parent frame

--- a/src/seahawk_rov/seahawk_rov/__main__.py
+++ b/src/seahawk_rov/seahawk_rov/__main__.py
@@ -64,12 +64,12 @@ def main(args=None):
 
     # instanciate the sensor classes
     logic_tube_bme280 = seahawk_rov.LogicTubeBME280(node_seahawk_rov, i2c)
-    # logic_tube_bno085 = LogicTubeBNO085(node_seahawk_rov, i2c)
+    logic_tube_bno085 = seahawk_rov.LogicTubeBNO085(node_seahawk_rov, i2c)
     thrust_box_bme280 = seahawk_rov.ThrustBoxBME280(node_seahawk_rov, i2c)
     
     def publisher():
         logic_tube_bme280.publish()
-        # logic_tube_bno085.publish()
+        logic_tube_bno085.publish()
         thrust_box_bme280.publish()
 
     publish_timer = node_seahawk_rov.create_timer(0.1, publisher)

--- a/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
+++ b/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
@@ -57,8 +57,6 @@ class LogicTubeBNO085:
         msg.linear_acceleration.x, msg.linear_acceleration.y, msg.linear_acceleration.z = self.bno.linear_acceleration
         msg.angular_velocity.x, msg.angular_velocity.y, msg.angular_velocity.z = self.bno.gyro
         msg.orientation.x, msg.orientation.y, msg.orientation.z, msg.orientation.w = self.bno.geomagnetic_quaternion
-        self.get_logger().info(f"Orientation  : {self.bno.geomagnetic_quaternion}")
-        self.get_logger().info(f"Acceleration : {self.bno.linear_acceleration}")
-        self.get_logger().info(f"Rotation     : {self.bno.gyro}")
+        
         # publish
         self.publisher.publish(msg)

--- a/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
+++ b/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
@@ -55,11 +55,18 @@ class LogicTubeBNO085:
         msg.header.frame_id = "logic_tube_bno085"
 
         # load the message with data from the sensor
-        # IMU X right, Y up, Z back
-        # ROV Y right, Z down, X forward
-        msg.orientation.x, msg.orientation.y, msg.orientation.z, msg.orientation.w = self.bno.geomagnetic_quaternion
-        msg.angular_velocity.x, msg.angular_velocity.y, msg.angular_velocity.z = self.bno.gyro
-        msg.linear_acceleration.x, msg.linear_acceleration.y, msg.linear_acceleration.z = self.bno.linear_acceleration
+        # IMU X right, Y forward, Z up
+        # ROV Y right, X forward, Z down
+        msg.orientation.x = self.bno.geomagnetic_quaternion[1]
+        msg.orientation.y = self.bno.geomagnetic_quaternion[0]
+        msg.orientation.z = -self.bno.geomagnetic_quaternion[2]
+        msg.orientation.w = self.bno.geomagnetic_quaternion[3]
+        msg.angular_velocity.x = self.bno.gyro[1]
+        msg.angular_velocity.y = self.bno.gyro[0]
+        msg.angular_velocity.z = -self.bno.gyro[2]
+        msg.linear_acceleration.x = self.bno.linear_acceleration[1]
+        msg.linear_acceleration.y = self.bno.linear_acceleration[0]
+        msg.linear_acceleration.z = -self.bno.linear_acceleration[2]
 
         # publish
         self.publisher.publish(msg)

--- a/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
+++ b/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
@@ -42,9 +42,9 @@ class LogicTubeBNO085:
         self.bno = BNO08X_I2C(i2c)
 
         # enable raw data outputs
-        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_RAW_ACCELEROMETER)
-        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_RAW_GYROSCOPE)
-        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_RAW_MAGNETOMETER)
+        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_LINEAR_ACCELERATION)
+        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_GYROSCOPE)
+        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_GEOMAGNETIC_ROTATION_VECTOR)
 
     def publish(self):
         # instantiate an imu message
@@ -54,9 +54,11 @@ class LogicTubeBNO085:
         msg.header.frame_id = "base_link"
 
         # load the message with data from the sensor
-        msg.linear_acceleration = self.bno.raw_acceleration
-        msg.angular_velocity = self.bno.raw_gyro
-        msg.orientation = self.bno.raw_quaternion
-
+        # msg.linear_acceleration = self.bno.linear_acceleration
+        # msg.angular_velocity = self.bno.gyro
+        # msg.orientation = self.bno.geomagnetic_quaternion
+        self.get_logger().info(f"Orientation  : {self.bno.geomagnetic_quaternion}")
+        self.get_logger().info(f"Acceleration : {self.bno.linear_acceleration}")
+        self.get_logger().info(f"Rotation     : {self.bno.gyro}")
         # publish
         self.publisher.publish(msg)

--- a/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
+++ b/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
@@ -29,22 +29,23 @@ import time
 # ros message
 from sensor_msgs.msg import Imu
 
-# inport the bno085 circuit python sensor library
+# import the bno085 circuit python sensor library
 import adafruit_bno08x
 from adafruit_bno08x.i2c import BNO08X_I2C
+
 
 class LogicTubeBNO085:
     def __init__(self, node, i2c):
         # instantiate the publisher
         self.publisher = node.create_publisher(Imu, 'logic_tube/imu', 8)
 
-        # instanciate the sensor
+        # instantiate the sensor
         self.bno = BNO08X_I2C(i2c)
 
         # enable raw data outputs
-        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_LINEAR_ACCELERATION)
-        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_GYROSCOPE)
         self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_GEOMAGNETIC_ROTATION_VECTOR)
+        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_GYROSCOPE)
+        self.bno.enable_feature(adafruit_bno08x.BNO_REPORT_LINEAR_ACCELERATION)
 
     def publish(self):
         # instantiate an imu message
@@ -54,9 +55,9 @@ class LogicTubeBNO085:
         msg.header.frame_id = "base_link"
 
         # load the message with data from the sensor
-        msg.linear_acceleration.x, msg.linear_acceleration.y, msg.linear_acceleration.z = self.bno.linear_acceleration
-        msg.angular_velocity.x, msg.angular_velocity.y, msg.angular_velocity.z = self.bno.gyro
         msg.orientation.x, msg.orientation.y, msg.orientation.z, msg.orientation.w = self.bno.geomagnetic_quaternion
-        
+        msg.angular_velocity.x, msg.angular_velocity.y, msg.angular_velocity.z = self.bno.gyro
+        msg.linear_acceleration.x, msg.linear_acceleration.y, msg.linear_acceleration.z = self.bno.linear_acceleration
+
         # publish
         self.publisher.publish(msg)

--- a/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
+++ b/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
@@ -52,9 +52,11 @@ class LogicTubeBNO085:
         msg = Imu()
 
         # add the frame id
-        msg.header.frame_id = "base_link"
+        msg.header.frame_id = "logic_tube_bno085"
 
         # load the message with data from the sensor
+        # IMU X right, Y up, Z back
+        # ROV Y right, Z down, X forward
         msg.orientation.x, msg.orientation.y, msg.orientation.z, msg.orientation.w = self.bno.geomagnetic_quaternion
         msg.angular_velocity.x, msg.angular_velocity.y, msg.angular_velocity.z = self.bno.gyro
         msg.linear_acceleration.x, msg.linear_acceleration.y, msg.linear_acceleration.z = self.bno.linear_acceleration

--- a/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
+++ b/src/seahawk_rov/seahawk_rov/logic_tube_bno085.py
@@ -54,9 +54,9 @@ class LogicTubeBNO085:
         msg.header.frame_id = "base_link"
 
         # load the message with data from the sensor
-        # msg.linear_acceleration = self.bno.linear_acceleration
-        # msg.angular_velocity = self.bno.gyro
-        # msg.orientation = self.bno.geomagnetic_quaternion
+        msg.linear_acceleration.x, msg.linear_acceleration.y, msg.linear_acceleration.z = self.bno.linear_acceleration
+        msg.angular_velocity.x, msg.angular_velocity.y, msg.angular_velocity.z = self.bno.gyro
+        msg.orientation.x, msg.orientation.y, msg.orientation.z, msg.orientation.w = self.bno.geomagnetic_quaternion
         self.get_logger().info(f"Orientation  : {self.bno.geomagnetic_quaternion}")
         self.get_logger().info(f"Acceleration : {self.bno.linear_acceleration}")
         self.get_logger().info(f"Rotation     : {self.bno.gyro}")

